aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -750,41 +749,55 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
   new_sizes.erase(new_sizes.begin() + dim);
   auto options = self.options();
 
-  // Selecting batch dimension
-  if (dim < self.dim() - 2) {
-    if (self.layout() == kSparseBsr) {
-      return at::native::_sparse_bsr_tensor_unsafe(
-          self.crow_indices().select(dim, index),
-          self.col_indices().select(dim, index),
-          self.values().select(dim, index),
-          new_sizes,
-          optTypeMetaToScalarType(options.dtype_opt()),
-          options.layout_opt(),
-          options.device_opt(),
-          options.pinned_memory_opt());
-    }
-    return at::native::_sparse_csr_tensor_unsafe(
-        self.crow_indices().select(dim, index),
-        self.col_indices().select(dim, index),
+  Tensor plain_indices;
+  Tensor compressed_indices;
+  std::tie(compressed_indices, plain_indices) =
+      AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+          self.layout(),
+          "select",
+          [&]() {
+            return std::make_pair(self.crow_indices(), self.col_indices());
+          },
+          [&]() {
+            return std::make_pair(self.ccol_indices(), self.row_indices());
+          });
+  auto n_batch = compressed_indices.dim() - 1;
+
+  if (dim < n_batch) {
+    // Selecting batch dimension
+    return at::native::_sparse_compressed_tensor_unsafe(
+        compressed_indices.select(dim, index),
+        plain_indices.select(dim, index),
         self.values().select(dim, index),
         new_sizes,
         optTypeMetaToScalarType(options.dtype_opt()),
         options.layout_opt(),
         options.device_opt(),
         options.pinned_memory_opt());
-  } else {
+  } else if (dim < n_batch + 2) {
+    // Selecting sparse dimension
     TORCH_CHECK(
-        self.is_sparse_csr(),
-        "select(): selecting non-batch dimensions is currently only supported for CSR tensors.");
+        self.layout() == kSparseCsr || self.layout() == kSparseCsc,
+        "select(): selecting non-batch dimensions is currently only supported for non-blocked sparse compressed layouts tensors.");
     TORCH_CHECK(
         self.dim() == 2,
```

Comment:
I think this check is not needed anymore, no?

Reply:
Ah, sorry, missed you comment from above about `to_sparse` working only for 2d. Is that really???

Reply:
I just tested and it appears that this is not completely correct. The call will throw if indices are not vectors specifically the compressed indicates this is due to missing support in `_convert_indices_from_csr_to_coo`, so dense dims are supported, but not batch dims. I will adjust the check, the error is still opaque to the user if we just allow the inner check to throw.